enhancements/ingress/bind-options.md

```diff
@@ -0,0 +1,268 @@
+
+```yaml
+---
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: default-1
+  namespace: openshift-ingress-operator
+spec:
+  endpointPublishingStrategy:
+    type: HostNetwork
+---
+apiVersion: operator.openshift.io/v1
+kind: IngressController
+metadata:
+  name: default-2
+  namespace: openshift-ingress-operator
+spec:
+  endpointPublishingStrategy:
+    type: HostNetwork
+    hostNetwork:
+      bindOptions:
+        httpPort: 11080
+        httpsPort: 11443
+        sniPort: 11444
+        noSniPort: 11445
+        statsPort: 1937
+```
+
+### Validation
+
+Specifying `spec.endpointPublishingStrategy.type: HostNetwork` and omitting
+`spec.endpointPublishingStrategy.hostNetwork` or
+`spec.endpointPublishingStrategy.hostNetwork.bindOptions` is valid and specifies
+the default behavior, which is to use the default ports to bind to. The API validates
+that any value provided for the ports defined in
+`spec.endpointPublishingStrategy.hostNetwork.bindOptions` is greater than `1`, and lower than `30000` (to avoid conflicts with 30000-32767 nodePort services range).
+
+
+### User Stories
+
+#### As a cluster administrator, I need to configure bind ports for my IngressController
+
+To satisfy this use-case, the cluster administrator can set the
+IngressController's `spec.endpointPublishingStrategy.hostNetwork.bindOptions`.
+
+
+### Implementation Details
+
+Implementing this enhancement requires changes in the following repositories:
+
+* openshift/api
+* openshift/cluster-ingress-operator
+
+OpenShift Cluster Ingress Operator, creates a deployment for Router with environment variables for port bindings which Openshift Router already respects. `ROUTER_SERVICE_HTTP_PORT`, `ROUTER_SERVICE_HTTPS_PORT`, `ROUTER_SERVICE_SNI_PORT`, `ROUTER_SERVICE_NO_SNI_PORT`, `STATS_PORT`.
+
+### Risks and Mitigations
+
+
+Users might define conflicting ports which causes haproxy process fail to start. A mitigation to this risk is to implement a validation feature in cluster-ingress-operator in reconsiler loop to ensure all the ports defined in the `bindOptions` section are unique, and return err if there are conflicting ports with a meaningful error message.
+
+Also if the underlying IngressController implementation were to change away from HAProxy to a different implementation, some values such as `sniPort` and `noSniPort` might not be needed and loose their meaning. Also the `statsPort` teminology might be different for other tools such as `Envoy`, which `metricPort` might be more meaningful in that case.
+
+
+## Design Details
+
+### Test Plan
+
+The controller that manages the IngressController Deployment and related
+resources has unit-test coverage; for this enhancement, the unit tests are
+expanded to cover the additional functionality.
+
+The operator has end-to-end tests; for this enhancement, the following test can
+be added:
```

Comment:
What you described below is actually the unit test that you did. The e2e tests typically involve the deployment of the ingress controllers on the real Kubernetes/OpenShift cluster.

`cluster-ingress-operator` repository has a dedicated directory for e2e tests and it already has a scenario for the host network type of ingress controller: [here](https://github.com/openshift/cluster-ingress-operator/blob/master/test/e2e/operator_test.go#L727). I was thinking that this e2e test case can be updated or a new one can be created to try a scenario similar to this:
- deploy host network ingress controller with the default port bindings
- deploy another host network controller with the same node placements but with `bindOptions` field set with the ports not conflicting with the first ingress controller
- check that the second ingress controller becomes ready
- check that the first ingress controller is still ready

Reply:
Moved them under the unit tests and added the mentioned scenario for e2e test.